File: Reqnroll/Chapter17/17.2-InPlacePersonaPattern/After/WIMP.Specs/Support/CustomParameterTypes.cs
using Reqnroll;

namespace WIMP.Specs.Support;

[Binding]
public class CustomParameterTypes(OrderingContext orderingContext)
{
    [StepArgumentTransformation(@"(\d+-\d+-\d+)")]
    public DateOnly ConvertDateOnly(string value)
    {
        return DateOnly.Parse(value);
    }

    [StepArgumentTransformation("the order|the placed order", Name = "order")]
    public int ConvertOrder()
    {
        return orderingContext.CurrentOrderNo ??
            throw new InvalidOperationException("No current order");
    }

    [StepArgumentTransformation(@"order ([A-Z])", Name = "order")]
    public int ConvertNamedOrderNumber(string name)
    {
        return orderingContext.NamedOrders.TryGetValue(name, out int orderNo)
            ? orderNo
            : throw new InvalidOperationException($"Order {name} not known");
    }
}

File: CucumberJVM/Chapter17/17.2-InPlacePersonaPattern/After/src/test/java/com/wimp/app/specs/support/CustomParameterTypes.java
package com.wimp.app.specs.support;

import com.wimp.app.models.OrderStatus;
import io.cucumber.java.DataTableType;
import io.cucumber.java.ParameterType;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.Map;

public class CustomParameterTypes {

    private final OrderingContext orderingContext;

    public CustomParameterTypes(OrderingContext orderingContext) {
        this.orderingContext = orderingContext;
    }

    @ParameterType("the order|the placed order|order [A-Z]")
    public int order(String value) {
        if (value.matches("the order|the placed order")){
            return orderingContext.getCurrentOrderNoVerified();
        }
        if (value.matches("order [A-Z]")){
            String name = value.substring(value.length() - 1);
            Integer orderNo = orderingContext.getNamedOrders().get(name);
            if (orderNo == null) {
                throw new RuntimeException("Order " + name + " not known");
            }
            return orderNo;
        }
        throw new RuntimeException("Invalid order value: " + value);
    }

    @DataTableType
    public NamedOrderData namedOrderDataRow(Map<String, String> row) {
        var orderData = new NamedOrderData();
        if (row.containsKey("order name"))
            orderData.setOrderName(row.get("order name"));
        if (row.containsKey("status"))
            orderData.setStatus(OrderStatus.valueOf(row.get("status").toUpperCase()));
        return orderData;
    }

    @ParameterType("\\d{4}-\\d{2}-\\d{2}")
    public LocalDate date(String value) {
        return LocalDate.parse(value);
    }

    @ParameterType("\\$?(\\d+(?:\\.\\d{1,2})?)")
    public BigDecimal price(String value){
        return new BigDecimal(value);
    }
}
